tools/generators/ethereum/contract_events.go.tmpl

```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
-    go func() {
-    	go watch() // trigger first subscription
-
-    	for {
-    		select {
-    		case <-errorChan:
-    		    {{$logger}}.Warning(
-                    "subscription to event {{$event.CapsName}} terminated with error; " +
-                        "resubscription attempt will be performed after the retry delay",
-                )
-    			time.Sleep(retryDelay)
-    			go watch()
-    		case <-unsubscribeChan:
-    			// shutdown the resubscriber goroutine on unsubscribe signal
-    			return
-    		}
-    	}
-    }()
-
-    // closing the unsubscribeChan will trigger a unsubscribe signal and
-    // run unsubscription for all subscription instances
-    unsubscribeCallback := func() {
-    	close(unsubscribeChan)
-    }
-
-    return subscription.NewEventSubscription(unsubscribeCallback), nil
-}
-
-func ({{$contract.ShortVar}} *{{$contract.Class}}) subscribe{{$event.CapsName}}(
-	success {{$contract.FullVar}}{{$event.CapsName}}Func,
-	fail func(err error) error,
-	{{$event.IndexedFilterDeclarations -}}
-) (subscription.EventSubscription, error) {
-	eventChan := make(chan *abi.{{$contract.AbiClass}}{{$event.CapsName}})
-	eventSubscription, err := {{$contract.ShortVar}}.contract.Watch{{$event.CapsName}}(
-		nil,
-		eventChan,
-		{{$event.IndexedFilters}}
-	)
-	if err != nil {
-		close(eventChan)
-		return eventSubscription, fmt.Errorf(
-			"error creating watch for {{$event.CapsName}} events: [%v]",
-			err,
-		)
-	}
-
-	var subscriptionMutex = &sync.Mutex{}
+	ctx, cancel := context.WithCancel(context.Background())
 
+	// TODO: Watch* function will soon accept channel as a parameter instead
+	// of the callback. This loop will be eliminated then.
 	go func() {
 		for {
 			select {
-			case event, subscribed := <-eventChan:
-				subscriptionMutex.Lock()
-				// if eventChan has been closed, it means we have unsubscribed
-				if !subscribed {
-					subscriptionMutex.Unlock()
-					return
-				}
+			case <-ctx.Done():
+				return
+			case event := <-eventOccurred:
 				success(
                     {{$event.ParamExtractors}}
 				)
-				subscriptionMutex.Unlock()
-			case ee := <-eventSubscription.Err():
-				fail(ee)
-				return
 			}
 		}
 	}()
 
-	unsubscribeCallback := func() {
-		subscriptionMutex.Lock()
-		defer subscriptionMutex.Unlock()
-
-		eventSubscription.Unsubscribe()
-		close(eventChan)
+	subscribeFn := func(ctx context.Context) (event.Subscription, error) {
+		return {{$contract.ShortVar}}.contract.Watch{{$event.CapsName}}(
+			&bind.WatchOpts{Context: ctx},
+			eventOccurred,
+			{{$event.IndexedFilters}}
+		)
 	}
 
-	return subscription.NewEventSubscription(unsubscribeCallback), nil
+	sub := ethutil.WithResubscription(
+		{{$contract.ShortVar}}SubscriptionBackoffMax,
+		subscribeFn,
+		{{$contract.ShortVar}}SubscriptionAlertThreshold,
+		func(elapsed time.Duration) {
+			{{$logger}}.Errorf(
+					"subscription to event {{$event.CapsName}} had to be "+
+						"retried [%v] since the last attempt; please inspect "+
```

Comment:
Will it print fine with `%v` shouldn't it be `%s`?
```suggestion
						"retried [%s] since the last attempt; please inspect "+
```

Reply:
It will work fine with both, I can replace `%v` with `%s` for clarity.

Reply:
See 64cab832ccbd7ae02b6e87da48ff65bf92fd8c1f please.